protocols/pop3/src/main/java/org/apache/james/protocols/pop3/core/AbstractPOP3MessageCommandHandler.java

```diff
@@ -0,0 +1,84 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.protocols.pop3.core;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.apache.james.protocols.api.ProtocolSession;
+import org.apache.james.protocols.api.Request;
+import org.apache.james.protocols.api.Response;
+import org.apache.james.protocols.pop3.POP3Response;
+import org.apache.james.protocols.pop3.POP3Session;
+import org.apache.james.protocols.pop3.mailbox.MessageMetaData;
+
+import com.google.common.collect.ImmutableList;
+
+public abstract class AbstractPOP3MessageCommandHandler extends AbstractPOP3CommandHandler {
```

Comment:
An abstact class of an abstract class  seems a bit far fetched to me.

We used to / still have similar code in other places and this had never ended up  turning out great.

I am an advocate of composition compared to inheritence when it comes up to code sharing.

For instance we could have a cool little POJO helper being `MessageCommand`

```
// POJO
class MessageCommand { 
   static Optional<MessageCommand> fromLine(String line) { ... here goes the parsing logic not linked to the processor stuff...  } 
   
   private final int num;
}
```

Reply:
This would make it easier to write unit test!

Reply:
AbstractPOP3CommandHandler was only introduced to get a grip on POP3 with Glowroot, and it does not really have any code in it. Of course the AbstractPOP3MessageCommandHandler could be a utility class instead, that accepts one or two lambdas instead of having protected methods. I am not quite sure how your MessageCommand code example fits in here though.

Reply:
> I am not quite sure how your MessageCommand code example fits in here though.

To be fair it would just wrap the line parsing part.

Reply:
In the second commit I switched from inheritance to delegation. Is this closer to what you envisioned?